Browser history support

goto-bus-stop opened the pull request:
This patch replaces the `history` array and hash URLs with a `pushState` based solution. For clarity I'll refer to the `history` array that stored UI instances as `customHistory`.

Demo: https://upload.clinicjs.org/public/02047f8379fae585fcdc17a644648fcd765fa46aa20ddf355761dafaae8698ae/2575.clinic-bubbleprof.html

TODO:

 - [x] When flicking through the history very quickly, animations are not queued, so you may still end up with multiple layouts showing at once.
 - [x] Returning to a top level collapsed node (eg #x3) does not work
 - [x] Remove debug logging stuff

Details:

 - On load, we set the current history state to `{ initial: true }` using `replaceState`. We can then do `!window.history.state.initial` to check if we have any entries in the history, instead of checking `customHistory.length`, to decide if the back button should be shown.
 - BubbleprofUI instances now have a `getHash()` method that returns a URL-compatible ID, just like the `location.hash` assignments before
 - On Bubbleprof 'navigation' events, we call `getHash()` on the `to` argument, and push the hash to the history using `pushState()`.
 - On browser 'popstate' events, we
   1. clear all sublayouts using `traverseUp()`—to do this I moved the 'setTopmostUI' handler from the close button method to `setupHistory()`. It _could_ be in the constructor but that would prevent GC of all layouts I think? It's only used by the history code which is only run in the root/original UI.
   2. parse the hash just like the `location.hash` parsing did on load before, passing `{silent: true}` to the 'navigation' event to prevent it from doing `pushState()` again
   3. show or hide the back button depending on the `history.state.initial` value
 - when the back button is clicked, we now just use the `history.back()` browser API which will trigger 'popstate'.

Q:

- Does 'navigation' ever have a `to` argument that's not a BubbleprofUI instance? This patch assumes `to` is always a BubbleprofUI. I clicked around a lot so I don't think so, but maybe I missed a button.

Fixes #158

AlanSl commented on visualizer/draw/bubbleprof-ui.js:
```diff
@@ -568,24 +590,23 @@ class BubbleprofUI extends EventEmitter {
+        // Close stack frames when moving away from their node.
+        if (this.topMostUI.selectedDataNode) {
+          this.topMostUI.clearFrames()
+        }
```
We've got a bug here - it doesn't handle sublayouts within sublayouts (ie. cluster nodes within collapsed nodes, or collapsed nodes within collapsed nodes). 

I think this slipped through your manual testing because by chance, you managed to generate a profile with no collapsed nodes! This is pretty rare, I'd be interested to see the application you used to generate it and add this sample to the sample set, it's pretty unique :-)

Here's an example that demonstrates the bug created using this PR: https://upload.clinicjs.org/public/07396a39c703c9ba755914d9e909033778049297e754f7f2767ff7b9b3d7cc66/11082.clinic-bubbleprof.html - click into the main MongoDB bubble, then click in to the first node in the chain (name ends with "&..."), then use the back button.

In trying to understand why the bug happens, I think I found something that fixes it, but I've only done very basic testing on this code change, so experiment with it, it probably needs altering: 

```
    window.addEventListener('popstate', (event) => {
      const { hash } = event.state
      // Close stack frames when moving away from their node.
      if (this.topMostUI.selectedDataNode) {
        this.topMostUI.clearFrames()
      }

      if (hash) {
        this.topMostUI.parseHash(hash)
      } else {
        this.topMostUI.traverseUp(null, { silent: true })
      }
    })
```

Sample: https://upload.clinicjs.org/public/1324c29b45787850e89d4c81c7de223379c49f4fad86a94a94dc802ccff023d6/11082.clinic-bubbleprof.html

Changes explained:

- If we always use traverseUp, we're assuming we need to go to all the way to the main layout first, which isn't true if we're stepping back within nested sublayouts. The jumpToNode / selectNode chain called from parseHash should handle any steps back up the tree required, it's designed to handle and animate any jump from anywhere to anywhere (e.g. using the buttons in the stack frames or search feature), so we shouldn't need to worry about that here unless we know we're going _all_ the way back because it's an empty hash. 
- `parseHash` should be called from the topmostUI here. Jumps to collapsed nodes need awareness of the UI and lineage of parent UIs. This is a tricky thing to understand, I discussed it briefly earlier but it takes some time and some experimentation with practical examples to get used to it. For example, if there is a hash `#x2-x1-c2` (x = collapsed node, c = cluster node), that means "find x2 within x1 within c2"; c2 (cluster node 2) can be located from anywhere because cluster nodes are unique and entirely determined by the clusterNode data from the dataSet, but the x1 referred to in the hash can only be located from within the context of c2, because it was only while processing the c2 cluster node's sublayout that it was identified that some nodes should be collapsed together to form collapsed node x1; and the x2 in the hash can only be located from the context of the x1 within c2 for the same reason. So, if were are in `#x1-c2` and we were previously in `#x2-x1-c2`, handling a back button press needs to look for `x2` using the context of the UI and layout objects of `#x1-c2`.

goto-bus-stop replied:
Ahh, doing `topMostUI.parseHash()` instead of `originalUI.parseHash()` also fixes the problem I encountered earlier, where moving to a different layout using these arrows would cause both layouts to show at once:
![image](https://user-images.githubusercontent.com/1006268/44903145-0c289880-ad0c-11e8-98d6-368076c64ebb.png)
Calling `traverseUp()` first fixed that as well but as you point out that's not quite correct.

With only this change, moving back from a hash like `#x1-x1-c3` to `#x1-c3` shows both the x1-c3 and c3 layouts at the same time. Time to test if my understanding of the layouts is now sufficient to debug this :)

goto-bus-stop replied:
Ok, using an animationQueue array in `parseCollapsedHash` fixed it!

goto-bus-stop commented:
Took a while but I found the problem that was causing jumping between `#x1-x3` and `#x3` hashes to bug out. Now, like doing `jumpToNode()` on a parent of the current top-most layout, moving from `#x1-x3` to `#x3` first zooms out fully and then zooms back in.

That is not the best UX but as it's in line with the `jumpToNode()` behaviour it might be best addressed in a separate PR: instead of simply walking up the ancestry to the original UI, we'd check each step of the way if it is the collapsed node we're looking for (by doing `parentUI.layout.layoutNodes.has(id)`, I guess?). Then we can have a similarly smooth experience there as we have with the breadcrumbs that do `traverseUp()`. At least, that's how it might be done based on my current understanding of bubbleprof :)

The history queueing problem still exists, I'll try the mid-animation queue push thing I asked about on slack after today's call.

AlanSl commented:
I was actually reviewing this and just about to say that it looks like the main problem now is animation chaining when jumping to an arbitrary collapsed node, which is a problem probably best addressed outside of this PR. It's likely to be a tricky one because it would require some kind of deeper seperation of creating layout objects and drawing uis which could involve tweaking a lot of things. 

What you're suggesting is fine since it should only apply in cases like someone manually typing a hash or using the back button after doing a deep jump like using a "select on diagram" button, and both are cases where seeing the new location relative to the main diagram rather than the previous layout makes sense.

goto-bus-stop commented:
I think I fixed the history animations bug! Just queueing the animations worked on the animation end but it didn't fix the problem, I think because some state updates are done synchronously (like `selectedDataNode`) and queueing animations across many sublayouts did not interact with that well.

This solution still uses an animation queue for the history, but instead of queueing more and more animations when flicking through the history quickly, it waits for the current animation to complete and then does a new `jumpToNode()` call. It only does that call for the most recent history update, so flicking through eg. 6 history entries very quickly will only animate from the current layout to the 1st entry, and then animate from the 1st entry straight to the 6th. This is actually pretty good because it means we don't spend 10 seconds catching up to the user's action.

A sample generated with this patch: https://upload.clinicjs.org/public/5ceb382664643e93dbf382d04f9e67510a968554ad9108cd9cea8cfa2d391707/10456.clinic-bubbleprof.html

I moved the animation queue stuff into a class mostly for debugging, so I could give queues a name and log the steps. It could be reverted back to an array if preferred but I don't think there are any downsides to this move.

goto-bus-stop commented:
New sample with a fix for history traversal after a queued jump was completely executed: https://upload.clinicjs.org/public/a92f085bc7b37c9c6df73e7d53a211d9975d994bf6e6c697032feb443bb36591/10456.clinic-bubbleprof.html

goto-bus-stop commented on visualizer/draw/bubbleprof-ui.js:
```diff
@@ -353,12 +329,81 @@ class BubbleprofUI extends EventEmitter {
+    })
+
+    window.addEventListener('popstate', (event) => {
+      const { hash } = event.state
```
Hmm now that you mention the error I remember I got that when manually editing the hash. Did you also manually edit the hash or did it happen during standard browser history traversal?

AlanSl replied:
I don't think so, most times I changed the hash nothing seemed to happen. Not impossible because I had earlier tried changing the hash manually in that profile.

goto-bus-stop commented on visualizer/draw/bubbleprof-ui.js:
```diff
@@ -353,12 +329,81 @@ class BubbleprofUI extends EventEmitter {
+      // If we have don't have a history animation queue, create one and jump
+      // immediately.
+      if (!this.historyAnimationQueue) {
+        jump()
```
That is when a non-history animation is going on, correct? Eg. clicking into a cluster node and then using the history before the animation completes.
I punted on that because it seemed a bit more tricky to handle non-history animations, since they can be started from any sublayout as well as the original UI. But probably the `historyAnimationQueue` could be replaced with a `this.originalUI.currentAnimationQueue` property and all UIs could use that in the same way to queue animations globally, so it might not be as difficult as I thought.

AlanSl replied:
Yes, that's right. Sounds like a good approach, treating animations the same as much as possible regardless of origin

AlanSl commented:
LGTM. I've tried to catch it out with some really complex queues but can't, seems really robust, code changes all make sense.
